Add ssl support

woile opened the pull request:
Finished work from songww in PR #213 

I tried to keep it faithful to songww code + added documentation and the url is properly displayed with https now.

songww committed:
Add ssl support.

woile committed:
docs: https information added

lovelydinosaur commented on .gitignore:
```diff
@@ -5,3 +5,10 @@ uvicorn.egg-info/
+
+# editors
+.vscode/
+.idea/
```
Editor ignores should stay in folks user specific gitignores.
The mypy one can go below .coverage without needing its own comment.

woile replied:
Oh didn't know about that, nice to learn something new. I'll update the PR

lovelydinosaur reviewed (changes requested):
Looking ace!
Ideally, UvicornWorker.run should pass through the Gunicorn ssl config options too.

woile committed:
feat: support for https in gunicorn worker

based on: https://github.com/benoitc/gunicorn/blob/33025cf610bb7a6f1cb307644c1881863c2fddc4/gunicorn/workers/gtornado.py#L137

woile commented:
Added ssl for gunicorn worker based on: https://github.com/benoitc/gunicorn/blob/33025cf610bb7a6f1cb307644c1881863c2fddc4/gunicorn/workers/gtornado.py#L137

lovelydinosaur commented on uvicorn/workers.py:
```diff
@@ -23,6 +24,12 @@ def run(self):
         }
+
+        if self.cfg.is_ssl:
+            _ssl_opt = copy.deepcopy(self.cfg.ssl_options)
```
I'd probably rather we explicitly copy across the options we need, rather than using `deepcopy`.

woile replied:
gonna fix this

woile replied:
I just made it a dictionary. Just for reference this is how `self.cfg.ssl_options` looks:
```python
{
    "keyfile": "./key.pem",
    "certfile": "./cert.pem",
    "ssl_version": 2,
    "cert_reqs": 0,
    "ca_certs": None,
    "suppress_ragged_eofs": True,
    "do_handshake_on_connect": False,
    "ciphers": "TLSv1",
}
```

lovelydinosaur commented on docs/deployment.md:
```diff
@@ -56,6 +58,15 @@ Options:
+                                  stdlib ssl module's)  [default: 0]
+  --ca-certs TEXT                 CA certificates file
+  --ciphers TEXT                  Ciphers to use (see stdlib ssl module's)
+                                  [default: TLSv1]
```
I'm thinking it might make sense for us to prefix all our command line options here as "ssl-...".

woile replied:
The only downside about doing this, is that it's no longer "compatible" with gunicorn.
Right now this way `uvicorn` and `gunicorn` share the same interface. So if you want to use ssl for development, and then go to production with gunicorn, your settings don't change. I think it might make it a bit annoying for uvicorn users. What do you think?

lovelydinosaur replied:
That's a good point.
I *think* I'm okay with that - we already differ on `--bind` vs. `--host`/`--port`.

lovelydinosaur commented on uvicorn/config.py:
```diff
@@ -72,6 +84,12 @@ def __init__(
+        ssl_version=ssl.PROTOCOL_TLS,
+        cert_reqs=ssl.CERT_NONE,
+        ca_certs=None,
+        ciphers="TLSv1",
```
Again, I think it might be a good idea for us to prefix all SSL options with "ssl_"

lovelydinosaur reviewed (changes requested):
Looking good - few more minor comments.
(Also I'm happy to pick up the last bits of this, if that helps)

woile commented:
@tomchristie a minor note, not related to this ticket per se, but maybe you have some ideas about it. 
I've noticied that I had to repeat the default values in the `config.Config` and in the click decorators. 
This kind of makes sense, because you might want to use uvicorn as a command line or as code. But if someone makes a mistake, there might be 2 different default settings. Is it okay? Should we care? How can I improve this?

lovelydinosaur commented:
> I've noticed that I had to repeat the default values in the `config.Config` and in the click decorators.

I think we just have to live with that yeah.

woile commented:
Oki, I'll leave it up to you then the decision to change the arguments to have a `ssl` prefix.

woile commented:
@tomchristie I see it reads "Changes requested" but I don't know what else needs to be updated, it shows nothing, what am I missing?

lovelydinosaur commented:
Nah, that's just because I hadn't got around to okay'ing it since your last update.

I *would* like to see "ssl_" argument prefixes and "ssl-" command line prefixes, but I'm perfectly happy to pick that bit up myself.

woile commented:
Ok, feel free to pick it up, otherwise I'll try to update the ssl prefix later tonight or tomorrow

woile committed:
feat: added prefix ssl to variables

woile committed:
test(ssl): new test case for https

woile commented:
Okay done, renamed variables to `ssl` and added a test :tada:

Commit:
Update settings.md

lovelydinosaur commented:
Amazing work here! I’ll plan to merge and release tomorrow.

lovelydinosaur commented:
Released to PyPI as 0.4.3
